airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftStagingS3Destination.java

```diff
@@ -60,44 +61,49 @@ private boolean isEphemeralKeysAndPurgingStagingData(final JsonNode config, fina
-          .withMessage(message);
-    } catch (final Exception e) {
-      LOGGER.error("Exception while checking connection: ", e);
-      return new AirbyteConnectionStatus()
-          .withStatus(AirbyteConnectionStatus.Status.FAILED)
-          .withMessage("Could not connect with provided configuration. \n" + e.getMessage());
-    } finally {
+      final S3DestinationConfig s3Config = getS3DestinationConfig(findS3Options(config));
+      final EncryptionConfig encryptionConfig =
+          config.has(UPLOADING_METHOD) ? EncryptionConfig.fromJson(config.get(UPLOADING_METHOD).get(JdbcUtils.ENCRYPTION_KEY)) : new NoEncryption();
+      if (isEphemeralKeysAndPurgingStagingData(config, encryptionConfig)) {
+        return new AirbyteConnectionStatus()
+            .withStatus(Status.FAILED)
+            .withMessage(
+                "You cannot use ephemeral keys and disable purging your staging data. This would produce S3 objects that you cannot decrypt.");
+      }
+      S3BaseChecks.attemptS3WriteAndDelete(new S3StorageOperations(new RedshiftSQLNameTransformer(), s3Config.getS3Client(), s3Config), s3Config,
+          s3Config.getBucketPath());
+
+      final NamingConventionTransformer nameTransformer = getNamingResolver();
+      final RedshiftS3StagingSqlOperations redshiftS3StagingSqlOperations =
+          new RedshiftS3StagingSqlOperations(nameTransformer, s3Config.getS3Client(), s3Config, encryptionConfig);
+      final DataSource dataSource = getDataSource(config);
       try {
-        DataSourceFactory.close(dataSource);
+        final JdbcDatabase database = new DefaultJdbcDatabase(dataSource);
+        final String outputSchema = super.getNamingResolver().getIdentifier(config.get(JdbcUtils.SCHEMA_KEY).asText());
+        attemptSQLCreateAndDropTableOperations(outputSchema, database, nameTransformer, redshiftS3StagingSqlOperations);
+        return new AirbyteConnectionStatus().withStatus(AirbyteConnectionStatus.Status.SUCCEEDED);
+      } catch (final ConnectionErrorException e) {
+        final String message = getErrorMessage(e.getStateCode(), e.getErrorCode(), e.getExceptionMessage(), e);
+        AirbyteTraceMessageUtility.emitConfigErrorTrace(e, message);
+        return new AirbyteConnectionStatus()
+            .withStatus(AirbyteConnectionStatus.Status.FAILED)
+            .withMessage(message);
       } catch (final Exception e) {
-        LOGGER.warn("Unable to close data source.", e);
+        LOGGER.error("Exception while checking connection: ", e);
+        return new AirbyteConnectionStatus()
+            .withStatus(AirbyteConnectionStatus.Status.FAILED)
+            .withMessage("Could not connect with provided configuration. \n" + e.getMessage());
+      } finally {
+        try {
+          DataSourceFactory.close(dataSource);
+        } catch (final Exception e) {
+          LOGGER.warn("Unable to close data source.", e);
+        }
       }
+    } catch (final Exception e) {
```

Comment:
@ryankfu does this match the semantics of a config error? Do we need to catch a more specific exception type?

my understanding is that we should only throw a config error if it's user error. I.e. if this is some sort of weird network thing (e.g. we messed up our GCP firewall rules and actually can't connect to S3) then we shouldn't wrap in a config error

* assuming we want to catch a more specific error type -> can we put this try-catch somewhere in base-java-s3? Maybe in S3BaseChecks or something. That would allow us to solve all S3-based destinations in one shot.
* or if we're ok catching generic Exception -> Can we just do that in IntegrationRunner? (https://github.com/airbytehq/airbyte/blob/3c52168f80cdc14cf346b4cadf5660fff0ad4936/airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java#L123)

Reply:
I didn't catch this tag before the changes were outdated but we don't need to catch a specific exception type so as long as the general exception type is directly associated with user-related errors

As you highlighted, firewall or transient issues _should not_ be captured as `ConfigErrorException`s